master/pkg/searcher/asha_stopping.go

```diff
@@ -0,0 +1,259 @@
+	// guarantee at least one trial at the top rung.
+	var ops []Operation
+	var maxConcurrentTrials int
+
+	if s.MaxConcurrentTrials > 0 {
+		maxConcurrentTrials = min(s.MaxConcurrentTrials, s.MaxTrials)
+	} else {
+		maxConcurrentTrials = max(
+			min(int(math.Pow(s.Divisor, float64(s.NumRungs-1))), s.MaxTrials),
+			1)
+	}
+
+	for trial := 0; trial < maxConcurrentTrials; trial++ {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		s.TrialRungs[create.RequestID] = 0
+		ops = append(ops, create)
+		ops = append(ops, NewTrain(create.RequestID, s.Rungs[0].UnitsNeeded))
+		ops = append(ops, NewValidate(create.RequestID))
+	}
+	return ops, nil
+}
+
+func (s *asyncHalvingStoppingSearch) trialCreated(
+	ctx context, requestID model.RequestID) ([]Operation, error) {
+	s.Rungs[0].OutstandingTrials++
+	s.TrialRungs[requestID] = 0
+	return nil, nil
+}
+
+func (s *asyncHalvingStoppingSearch) trialClosed(
+	ctx context, requestID model.RequestID) ([]Operation, error) {
+	s.TrialsCompleted++
+	s.ClosedTrials[requestID] = true
+	return nil, nil
+}
+
+func (s *asyncHalvingStoppingSearch) validationCompleted(
+	ctx context, requestID model.RequestID, validate Validate, metrics workload.ValidationMetrics,
+) ([]Operation, error) {
+	// Extract the relevant metric as a float.
+	metric, err := metrics.Metric(s.Metric)
+	if err != nil {
+		return nil, err
+	}
+	if !s.SmallerIsBetter {
+		metric *= -1
+	}
+
+	return s.promoteAsync(ctx, requestID, metric), nil
+}
+
+func (s *asyncHalvingStoppingSearch) promoteAsync(
+	ctx context, requestID model.RequestID, metric float64,
+) []Operation {
+	// Upon a validation complete, we should return at least one more train&val workload
+	// unless the bracket of successive halving is finished.
+	rungIndex := s.TrialRungs[requestID]
+	rung := s.Rungs[rungIndex]
+	rung.OutstandingTrials--
+	addedTrainWorkload := false
+
+	var ops []Operation
+	// If the trial has completed the top rung's validation, close the trial.
+	if rungIndex == s.NumRungs-1 {
+		rung.Metrics = append(rung.Metrics,
+			trialMetric{
+				RequestID: requestID,
+				Metric:    metric,
+			},
+		)
+
+		if !s.EarlyExitTrials[requestID] {
+			ops = append(ops, NewClose(requestID))
+			s.ClosedTrials[requestID] = true
+		}
+	} else {
+		// This is not the top rung, so do promotions to the next rung.
+		nextRung := s.Rungs[rungIndex+1]
+		promoteTrial := rung.continueTraining(
+			requestID,
+			metric,
+			s.Divisor,
+		)
+		if promoteTrial {
+			s.TrialRungs[requestID] = rungIndex + 1
+			nextRung.OutstandingTrials++
+			if !s.EarlyExitTrials[requestID] {
+				unitsNeeded := max(nextRung.UnitsNeeded.Units-rung.UnitsNeeded.Units, 1)
+				ops = append(ops, NewTrain(requestID, model.NewLength(s.Unit(), unitsNeeded)))
+				ops = append(ops, NewValidate(requestID))
+				addedTrainWorkload = true
+			} else {
+				// We make a recursive call that will behave the same
```

Comment:
thinking through this, I'm not sure if this code path is even possible anymore, if a trial exited early, it would've completed no operations and `trialExitedEarly` would've been called for it, if it is here and has completed a validation.. well it wouldn't have gotten here if it exited early.

Reply:
good catch.  i've modified how we handle early exit trials for stopping to never promote.  this is okay for stopping since we are guaranteed to have a trial train for the max_length as long as there is a trial that is not exited early.  this is not okay for promotion-based asha since we need to fill in enough trials in the rungs to promote one to the max_length.